ops/relation.py

```diff
@@ -0,0 +1,548 @@
+
+        - :class:`ProviderAvailable`
+        - :class:`ProviderInvalid`
+        - :class:`ProviderUnready`
+
+        Args:
+            event: an event object. It is expected that the event object
+                contains a key `provider_data` whose value is all the data
+                forwarded by the :class:`Provider` object.
+        """
+        rdata = event.relation.data[event.app]
+        logger.debug("Got data from provider : {}".format(rdata))
+        provider_data = rdata.get('provider_data')
+        consumed = self.consumes
+        if provider_data:
+            data = json.loads(provider_data)
+            try:
+                provides = data['provides']
+            except KeyError:
+                # provider has not set any specification
+                # so no compatibility checks are done
+                # and no events are raised
+                logger.warning('Provider not specified')
+                return
+        else:
+            logger.debug('No provider data')
+            # provider has not given any information
+            # so provider will not be made available (as yet)
+            return
+
+        ready = data.get('ready')
+        if not ready:
+            self.on.unready.emit()
+            return
+
+        stored_id = self.stored.relation_id
+        rel_id = event.relation.id
+        check_single = ((stored_id is None) or (stored_id == rel_id))
+        if self.multi_mode or check_single:
+            requirements_met = self.meets_requirements(provides, consumed)
+        else:
+            return
+
+        if requirements_met:
+            logger.debug('Got compatible provider : {}'.format(provider_data))
+            if not self.multi_mode and not self.stored.relation_id:
+                self.stored.relation_id = rel_id
+                logger.debug('Saved relation id : {}'.format(rel_id))
+            self.on.available.emit(data)
+        else:
+            logger.error('Incompatible provider : Need {}, Got {}'.format(
+                consumed, provider_data))
+            self.on.invalid.emit(provides)
+
+    def on_provider_broken(self, event):
+        """Inform consumer charm that provider relation no longer exists
+
+        This method raises a :class:`ProviderBroken` event in respose to
+        a relation broken event.
+
+        Args:
+            event: an event object
+        """
+        logger.debug("Provider Broken : {}".format(event))
+        if not self.multi_mode:
+            self.stored.relation_id = None
+        self.on.broken.emit()
+
+    def validate_provider(self, event):
+        """Check provider and consumer compatibility
+
+        This method validates provider consumer compatibility using
+        data that is already available in the application relation
+        bucket.
+
+        Args:
+            event: an event object
+        """
+        logger.debug("Validating provider(s) : {}".format(event))
+        consumed = self.consumes
```

Comment:
Your definition of consumes allowed for multiple things that could be consumed (the example was 'mysql' and 'mariadb'.

However this code assumes consumed is exactly consumes and later on is passed as 'a single value key: value dictionary'.

Reply:
`consumed = self.consumes` is a dictionary with potentially multiple key value pairs. Each pair is passed one at a time to `is_compatible` by `meets_requirements`. I hope this clarifies the question.